fedimint-core/src/db/mod.rs

```diff
@@ -286,6 +286,115 @@ pub trait IDatabaseTransaction<'a>: 'a + Send {
     async fn set_tx_savepoint(&mut self);
 }
 
+/// `ISingleUseDatabaseTransaction` re-defines the functions from
+/// `IDatabaseTransaction` but does not consumed `self` when committing to the
+/// database. This allows for wrapper structs to more easily borrow
+/// `ISingleUseDatabaseTransaction` without needing to make additional
+/// allocations.
+#[async_trait]
+pub trait ISingleUseDatabaseTransaction<'a>: 'a + Send {
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>>;
+
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>>;
+
+    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>>;
+
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> PrefixStream<'_>;
+
+    async fn raw_remove_by_prefix(&mut self, key_prefix: &[u8]) -> Result<()>;
+
+    async fn commit_tx(&mut self) -> Result<()>;
+
+    async fn rollback_tx_to_savepoint(&mut self);
+
+    async fn set_tx_savepoint(&mut self);
+}
+
+/// Struct that implements `ISingleUseDatabaseTransaction` and can be wrapped
+/// easier in other structs since it does not consumed `self` by move.
+pub struct SingleUseDatabaseTransaction<'a, Tx: IDatabaseTransaction<'a> + Send>(
+    Option<Tx>,
+    &'a PhantomData<()>,
+);
+
+impl<'a, Tx: IDatabaseTransaction<'a> + Send> SingleUseDatabaseTransaction<'a, Tx> {
+    pub fn new(dbtx: Tx) -> SingleUseDatabaseTransaction<'a, Tx> {
+        SingleUseDatabaseTransaction(Some(dbtx), &PhantomData)
+    }
+}
+
+#[async_trait]
+impl<'a, Tx: IDatabaseTransaction<'a> + Send> ISingleUseDatabaseTransaction<'a>
+    for SingleUseDatabaseTransaction<'a, Tx>
+{
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: Vec<u8>) -> Result<Option<Vec<u8>>> {
+        match &mut self.0 {
+            Some(tx) => tx.raw_insert_bytes(key, value).await,
+            None => Err(anyhow::anyhow!(
+                "Cannot insert into already consumed transaction"
+            )),
```

Comment:
try `self.0.as_mut().context(...).func(....)`

context is from `anyhow::Context`

Reply:
Fixed